### ISSUE TITLE
tedlium3 recipe softmax layer

### ISSUE THREAD

**Issue by AlexandderGorodetski** [k2-fsa/icefall#776]
Hello,

Following is training command of tedlium3 recipe

./pruned_transducer_stateless/train.py \
  --world-size 4 \
  --num-epochs 30 \
  --start-epoch 0 \
  --exp-dir pruned_transducer_stateless/exp \
  --max-duration 300

I just wanted to verify that softmax layer of RNN transducer contains all English characters in this case. Is that correct?

Thanks,
AlexG.

**Comment by marcoyang1998** [comment]
I think the default recipe is using a BPE model. You may check `tokens.txt` to see what the modeling units look like.

**Comment by AlexandderGorodetski** [comment]
I do not see file tokens.txt in exp directory. Could you please let me know where I can find it.
I do see file tokens.txt in directory lang_bpe_500, but this directory is not passed as a parameter to train.py, so I do not see any option that train.py can get tokens.txt file form lang_bpe_500 directory.

**Comment by csukuangfj** [comment]
> I do not see file tokens.txt in exp directory. Could you please let me know where I can find it.
> 
> I do see file tokens.txt in directory lang_bpe_500, but this directory is not passed as a parameter to train.py, so I do not see any option that train.py can get tokens.txt file form lang_bpe_500 directory.

Please read the code again. It is a default parameter and you don't need to pass it explicitly.

**Comment by csukuangfj** [comment]
> so I do not see any option that train.py can get tokens.txt file form lang_bpe_500 directory

Please read it below if you missed it:
https://github.com/k2-fsa/icefall/blob/65d7192dca03ba21bff4270add3891c9730491a7/egs/tedlium3/ASR/pruned_transducer_stateless/train.py#L121

`tokens.txt` is generated from `bpe.mode`. Here in training, we use only `bpe.model`. `tokens.txt` contains only the mapping between `tokens` and their integer IDs. During training, we want to encode words into tokens and we can only do it with a `bpe.model`, not with `tokens.txt`.

https://github.com/k2-fsa/icefall/blob/65d7192dca03ba21bff4270add3891c9730491a7/egs/tedlium3/ASR/pruned_transducer_stateless/train.py#L620
`vocab_size` specifies the dim of `softmax`.
